packages/typescript-estree/src/convert.ts

```diff
@@ -1044,7 +1044,8 @@ export class Converter {
         }
 
         if (
-          result.key.type === AST_NODE_TYPES.Identifier &&
+          (result.key.type === AST_NODE_TYPES.Identifier ||
+            result.key.type === AST_NODE_TYPES.MemberExpression) &&
           node.questionToken
         ) {
           result.key.optional = true;
```

Comment:
Why add it to `key`? Let's add it too the property node like Babel does. Or better, for backward compatibility, to both.

Reply:
Also the expression does not necessarily has to be a member expression. Syntactically, TS allows any expression there, it only puts some limitations on its type. E.g. check out this valid (albeit useless) TypeScript code:
```ts
declare function f(): string;

class Foo {
    [f()]?() {
        return 1;
    }
}
```
[playground](https://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVIAoBKALngGcMYtUBzAbgFgAoFyKc8+AMRx3gDeLeCPgBtRMQC6AfmKDhopXAzIY+AIxNmSgL4t9zIA)

Reply:
@thorn0 (/cc @bradzacher ) I also wondered about that. However, there is no `optional` in `MethodDefinitionBase`. (https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/ts-estree/ts-estree.ts#L510-L519). Can we update it?

Reply:
Hmm..
Tracing it back. This pattern goes back a _long_ way.
Originally introduced in 2017 https://github.com/eslint/typescript-eslint-parser/pull/187
Why did it get put there instead of on the property itself? I don't know.

Digging into this more - we don't do a good job of standardised support for optional members.
I'd be happy if you standardised things, without doing a breaking change.
We can file an issue for the 3.0.0 release to remove the old weird cases.

```ts
const a = 'aa';
const b = 'bb';
const c = 'cc';
const d = 'dd';
const e = 'ee';
const f = 'ff';

abstract class Foo {
    a?: string;                 // ClassProperty is marked optional
    [a]?: string;               // nothing is marked optional, should be ClassProperty
    ['aaa']?: string;           // nothing is marked optional, should be ClassProperty
    b?() {}                     // Identifier is marked optional, should be MethodDefinition
    [b]?() {}                   // Identifier is marked optional, should be MethodDefinition
    ['bbb']?() {}               // nothing is marked optional, should be MethodDefinition
    c?(): void                  // Identifier is marked optional, should be MethodDefinition
    [c]?(): void                // Identifier is marked optional, should be MethodDefinition
    ['ccc']?(): void            // nothing is marked optional, should be MethodDefinition
    
    abstract d?: string;        // TSAbstractClassProperty is marked optional
    abstract [d]?: string;      // nothing is marked optional, should be TSAbstractClassProperty
    abstract ['ddd']?: string;  // TSAbstractClassProperty is marked optional
    abstract e?(): void         // Identifier is marked optional, should be TSAbstractMethodDefinition
    abstract [e]?(): void       // Identifier is marked optional, should be TSAbstractMethodDefinition
    abstract ['eee']?(): void   // nothing is marked optional, should be TSAbstractMethodDefinition

    constructor(
        private f?: string,     // Identifier is marked optional, should be that AS WELL AS TSParameterProperty
    ) { }
}

interface Bar {
    a?: string;                 // TSPropertySignature is marked optional
    [a]?: string;               // TSPropertySignature is marked optional
    ['aaa']?: string;           // TSPropertySignature is marked optional
    b?(): void                  // TSMethodSignature is marked optional
    [b]?(): void                // TSMethodSignature is marked optional
    ['bbb']?(): void            // TSMethodSignature is marked optional
}

type Baz1 = {
    [k in string]?: string      // TSMappedType is marked optional
}
type Baz2 = {
    [k in string]+?: string     // TSMappedType is marked optional = "+"
}
type Baz3 = {
    [k in string]-?: string     // TSMappedType is marked optional = "-"
}

function Buzz(
  a?: string,                   // Identifier is marked optional
): void {}
```
[astexplorer repl](https://astexplorer.net/#/gist/2dd68dc85bd93d379f12ff61a1d08305/1a24fa85854621008953ce51604c6833f9f5cdb1)
[ts playground repl](https://www.typescriptlang.org/play/#code/MYewdgzgLgBAhjAvDA5HOKDcAoUlYBGSqBBWu40MwxKww5eVAJrc845bAKa3fed8MAGa1hw8tjgFoAJzjBYwADZwIEGADEQIGAG9sMI-AD8ALhhyAlmADmmY46dOA9C5gBhVeoAKskAAO3LJQAJ4wVhoAtnCyANbcrIFQVuBwyobGANpwALrmllCyNvbOZUZuMGAgUAAWJRHRsQlJASlpygA0lrUgAK7KrAS8XmoQfoHBYZlGWWjoKPkW1nYO5RXu1XUNkTAx8Ykwyalg6d0QvQNDI97j-kEhoTMwBCYAFACU+gC+638bMAAksxuGAUsIrMFGntmodjh1zpdBi9eABZbh1EDMAAi3AhYCs7TAzyyBHynx+-3KlWBoPBkNk0P2LSObROZx6-WRwxg6MxOLxNkJJxJKFIZHJXz0vypAK29TsTNhrSJHIuXOuvIxvQF+OF4GewHeHwsADcQFZWLLHDSQWCrBCobtmXC2QjOVcUVr+bi9USScBJWaLVbrba6Q6GUqDir2V0Pdy0dqsb6hf7HHN6AwgzBzZaqZV5TsmjHWar4+rPTy+TrUwT08ZntI5ApYMwCitSmVKgAVADKAEEZEVW6NfPcpuFncqy3Gm8P5IoYFlmEtCsVVs5CzUFbZoyz4acK0jNf2hy3FGO7pNHvOL7A5uwOGvO2te4OF6PbhMHmF967yzvEcl24Y1g3zal3Fpe1HUZadS0PNUTy9M9P0UGsU0FesRUcZtgIfbgczzUMbSgu16SdEsDzdI9EQ1FCP3vDDdTTHDjDwxcHxQfgBCIkNjG3bZFXg6jyzoqteFQpjkxY7CDUNLhZD6RQQFkN5nkcAJilNOAoF4YQOyKEpulIoFyMjSiYQQmikPonk6l0mABz7GAAHUAFEABlPKclz+x8WI4CiDFgh-SdnilGBvmwaLsBsPTZGEBReAAIVifQm0Mjcu2tGB3zCx4+ysWxTigPpZF4ESALnDM8iyko1llfKJ0K4rSvKyqqOqjpRXQDAXyMzc-mam8wiKkrdI6-9Yx6xxXk+cCSLDdx+2Y8b2oq6bZ1m7IyTA3N+Ny981raybNqqmaj1FcVFn24iCxWvsTomsrzq6y70hi7BsDCIIYDSgAvABGYgDAzOIIjAdcSgG7Kt0e1E4ACIJmB7UI-ou7arui37UrgAGACZQZJCGbGhuxcgAanqxVTNWpGUbRjH3qx9JiAAIkp9mvtx-78YAZmJ8HIfJ2xcgAWhpvc6aehnEiZzqrNEuMOfF7nYuEPowEUE5-r6AGAfUow4ClkzcoBaCKLglnEIyE0DvzaUgA)


LMK if this makes sense

Reply:
Thanks, @bradzacher!! ( /cc @thorn0 )
Personally, I think that it is better to keep adding optional to key until releasing 3.0.
Because, in order for us to add `optional` to properties, we need to update some basic node type definitions(e.g. `MethodDefinitionBase` ). I am reluctant to do it with minor version upgrades ...

Reply:
> We can file an issue for the 3.0.0 release to remove the old weird cases.

:+1:

Reply:
> in order for us to add optional to properties, we need to update some basic node type definitions

It's totally okay as long as the property you add is optional. 

"Standardising things, without doing a breaking change" means:
- in cases where `optional` is added to the `key` in the current version, add it to both the key and the property/method
- in other cases, add it only to the property/method